### DISCUSSION TITLE
How can I store data in a file and extract in a matrix form later?

### DISCUSSION THREAD

**Discussion by ShaikhaTheGreen** [zwicker-group/py-pde#39] [Q&A] [answered]
Thank you for the magnificent work in py-pde!

I'm simulating the 1D wave equation (with a time function source) in py-pde, and I want to compare the results to another results I got from another python code output.
I want to know how can I use` FileStorage` to save the fields:
x, t , u 

- How can I store the data in npz/ h5py file?
- How can I extract the data in my other code?

In my other code, I'm saving my data in an npz file:
```
data_file = 'mydata'
np.savez(data_file, t_input=t, x_input=x, y_input=y, u_output=exact )
```
and then loading them to check/plot the values:

```
data = np.load(data_file+'.npz')
t_i, x_i, y_i, exact_i = data["t_input"], data["x_input"], data["y_input"], data["u_output"]
```

How can I do something similar in py-pde?

Thanks in advance!

This my code:
```

from math import sqrt,pi
import pde
from pde import WavePDE, CartesianGrid, ScalarField, MemoryStorage, FieldCollection, PDE, plot_kymograph, PlotTracker

#Set the Geometry
#Generate grid
grid = CartesianGrid( [[-1, 1]] , 256 )


#Initial condition
u = ScalarField.from_expression(grid, "sin(pi*x)", label="Field $u$" )
v = ScalarField(grid, 0, label="Field $v$")
initial_state = FieldCollection([u, v])


#Boundary condition
#specified within eq definition
#eq = pde.WavePDE(speed = 0.01, bc=[{"value": 0}, {"derivative": 0}])
speed = 0.01
frequency = 10
source = f"sin(2*pi*{frequency}*t)"
eq = PDE(
    {
        "u": f"v",
        "v": f"{speed}**2 * laplace(u) + {source}"
    }, bc=[{"value": 0}, {"derivative":0}]
)

# store intermediate information of the simulation
storage = MemoryStorage()

# solve the pde
result = eq.solve(initial_state, t_range = 1, tracker=[storage.tracker(0.01)])
    
#create a simulation movie
pde.movie(storage, filename="1D_wave_movie.mov")

# plot the resulting field
result.plot()

# visualize the results as a space-time plot
plot_kymograph(storage, scalar="norm_squared")
```

**Comment by david-zwicker** [comment, marked as answer]
If you want to store the results on the hard drive, it might be best to use a [`FileStorage`](https://py-pde.readthedocs.io/en/latest/packages/pde.storage.html#module-pde.storage.file) instead of the `MemoryStorage` in your code. py-pde will then store the results in an hdf5 file which you can read either by initializing a `pde.FileStorage` with it or by just reading it using `h5py`. Note that the information about the grid (and thus the discretization) is stored in a hdf attribute and might be difficult to reconstruct manually. I thus suggest initializing the storage using `py-pde` again and then access the data using the `data` attribute. The underlying grid can be accessed using the `grid` attribute. In particular, [you get information about the discretization](https://py-pde.readthedocs.io/en/latest/packages/pde.grids.html#pde.grids.base.GridBase) using the `axes_coords` or `cell_coords` attributes.

**Reply by ShaikhaTheGreen** [reply]
Thanks, it worked!

I have noticed that the info recorded are only `data` and `times` as I checked for keys in the file. Is there a way to also record the values of x used?

**Reply by david-zwicker** [reply]
The discretization information is not written explicitly, but contained in an HDF attribute. When you open the data using `FileStorage`, the `grid` attribute gives you access to grid instance, from which you can read off the discretization (and other information).

**Comment by anilkunwar** [comment]
Thanks a lot for this information. The FileStorage class seems to be very useful for data storage in an external file.

The above code worked for me when I added  name of the file inside the brackets of pde.FileStorage()
`from pde import FileStorage`
...
```
# store intermediate information of the simulation
# storage = MemoryStorage()
storage = FileStorage(filename="a.hdf5")
```
